File: sdks/python/blocks_network/part_helpers.py
"""
Convenience helpers for building request parts.

Provides :func:`text_part` and :func:`file_part` so consumers can
construct :class:`SendMessageRequestPart` instances without knowing
wire format field names or manually reading files.
"""

from __future__ import annotations

from pathlib import Path
from typing import Optional, Union

from .task_client import SendMessageRequestPart


def text_part(text: str, part_id: str = "text") -> SendMessageRequestPart:
    """Build a text-only request part.

    Parameters
    ----------
    text:
        The text content.
    part_id:
        Part ID matching a declared ``io.inputs[].id`` in the agent
        card. Defaults to ``"text"``.
    """
    return SendMessageRequestPart(part_id=part_id, text=text)


def file_part(
    path_or_data: Union[str, Path, bytes, bytearray],
    *,
    part_id: str = "file",
    file_name: Optional[str] = None,
    content_type: str = "application/octet-stream",
) -> SendMessageRequestPart:
    """Build a file request part from a path or raw bytes.

    Parameters
    ----------
    path_or_data:
        File path (str or Path) to read, or raw bytes/bytearray.
    part_id:
        Part ID matching a declared ``io.inputs[].id`` in the agent
        card. Defaults to ``"file"``.
    file_name:
        Override file name. When reading from a path, defaults to
        the file's basename.
    content_type:
        MIME type. Defaults to ``"application/octet-stream"``.
    """
    if isinstance(path_or_data, (str, Path)):
        p = Path(path_or_data)
        return SendMessageRequestPart(
            part_id=part_id,
            file=p.read_bytes(),
            file_name=file_name or p.name,
            content_type=content_type,
        )
    return SendMessageRequestPart(
        part_id=part_id,
        file=bytes(path_or_data),
        file_name=file_name or "file",
        content_type=content_type,
    )

File: sdks/node/src/runtime/part-helpers.ts
/**
 * Part builder helpers for constructing SendMessageRequestPart objects.
 *
 * These simplify the consumer experience by handling wire format field
 * names and file reading automatically.
 *
 * Browser-safe: no top-level `node:fs` import. The Node-only
 * `filePartFromPath` helper lazy-imports `node:fs` on call so browser
 * bundlers can tree-shake it out of consumer bundles that don't use
 * filesystem paths.
 */

import type { SendMessageRequestPart } from './task-client.js';

/** Extract the base name from a file path (portable, no node:path needed). */
function baseName(filePath: string): string {
  const lastSlash = Math.max(filePath.lastIndexOf('/'), filePath.lastIndexOf('\\'));
  return lastSlash >= 0 ? filePath.slice(lastSlash + 1) : filePath;
}

/**
 * Build a text request part.
 *
 * @param text The text content.
 * @param partId The part ID (defaults to 'text'). Must match a declared
 *               io.inputs[].id in the agent's card.
 */
export function textPart(text: string, partId = 'text'): SendMessageRequestPart {
  return { partId, text };
}

/**
 * Build a file request part from raw file data.
 *
 * Universal: works in both Node and browser. No filesystem access.
 * For Node-only file-path convenience, use `filePartFromPath`.
 *
 * Accepts `Uint8Array`, `ArrayBuffer`, `Blob`, or `File`. Browser
 * consumers can pass a `File` from `<input type="file">` directly.
 *
 * @param data Raw file data.
 * @param options Optional overrides for partId, fileName, and contentType.
 */
export function filePart(
  data: Uint8Array | ArrayBuffer | Blob | File,
  options?: { partId?: string; fileName?: string; contentType?: string },
): SendMessageRequestPart {
  const partId = options?.partId ?? 'file';
  // Prefer an inferred fileName for File instances (browser convention).
  const inferredName =
    typeof File !== 'undefined' && data instanceof File ? data.name : undefined;
  const inferredType =
    typeof Blob !== 'undefined' && data instanceof Blob && data.type
      ? data.type
      : undefined;
  return {
    partId,
    file: data,
    fileName: options?.fileName ?? inferredName ?? 'file',
    contentType:
      options?.contentType ?? inferredType ?? 'application/octet-stream',
  };
}

/**
 * Build a file request part from a filesystem path (Node-only).
 *
 * Lazy-imports `node:fs` on first call so this helper can coexist
 * with browser bundlers without pulling `node:fs` into the bundle.
 * Not available in browsers.
 *
 * @param path Filesystem path.
 * @param options Optional overrides for partId, fileName, and contentType.
 */
export async function filePartFromPath(
  path: string,
  options?: { partId?: string; fileName?: string; contentType?: string },
): Promise<SendMessageRequestPart> {
  const { readFileSync } = await import('node:fs');
  const buffer = readFileSync(path);
  // Re-wrap as a fresh Uint8Array so downstream normalization treats
  // the data uniformly regardless of whether Buffer is defined.
  const data = new Uint8Array(buffer);
  return {
    partId: options?.partId ?? 'file',
    file: data,
    fileName: options?.fileName ?? baseName(path),
    contentType: options?.contentType ?? 'application/octet-stream',
  };
}
